cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py

```diff
@@ -17,18 +17,293 @@
+        if self.static_nodes > 0:
+            node_set += f"{self.static_node_name()},"
+        if self.dynamic_nodes > 0:
+            node_set += f"{self.dynamic_node_name()},"
+
+        return node_set
+
+    def as_gres(self):
+        """Alternative rendering for the gres config."""
+        gres_render = ""
+        if self.gpu_count > 0:
+            if self.static_nodes > 0:
+                gres_render += (
+                    f"NodeName={self.static_node_name()} Name=gpu "
+                    f"Type={ self.gpu_type } File=/dev/nvidia[0-{ self.gpu_count - 1 }]\n"
+                )
+            if self.dynamic_nodes > 0:
+                gres_render += (
+                    f"NodeName={self.dynamic_node_name()} Name=gpu "
+                    f"Type={ self.gpu_type } File=/dev/nvidia[0-{ self.gpu_count - 1 }]\n"
+                )
+
+        return gres_render
+
+    def _definitions(self, dynamic=False):
+        definitions = f" CPUs={ self.vcpus() } RealMemory={ self.real_memory } State=CLOUD {self._features(dynamic)}"
+
+        if self.has_gpu and self.gpu_count > 0:
+            definitions += f" Gres=gpu:{ self.gpu_type }:{self.gpu_count}"
+
+        return definitions
+
+    def _features(self, dynamic=False):
+        resource_type = "static"
+        if dynamic:
+            resource_type = "dynamic"
+
+        instance_type = f",{self.instance_types[0]}"
+        if len(self.instance_types) > 1:
+            # When multiple instance types are defined we do not know in advance which one will be used
+            # to launch the node. So we do not list any of them as feature
+            instance_type = ""
+
+        features = f"Feature={resource_type}{instance_type},{self.name}"
+        if self.efa_enabled:
+            features += ",efa"
+
+        if self.gpu_count > 0:
+            features += ",gpu"
+
+        return features
+
+    def _node_name(self, type, size):
+        return f"{self.queue_name}-{type}-{self.name}-[1-{ size }]"
+
+
+class QueueRenderer:
+    """Renders a PCluster Queue Config as a Slurm partition config or as gres config."""
+
+    def __init__(self, queue_config, no_gpu, memory_ratio, instance_types_info, conf_type="partition", default=False):
+        self.name = queue_config["Name"]
+        self.is_default = default
+        self.conf_type = conf_type
+        self.capacity_type = queue_config["CapacityType"]
+        self.iam = queue_config["Iam"]
+        self.compute_renderers = [
+            ComputeResourceRenderer(self.name, compute_resource_config, no_gpu, memory_ratio, instance_types_info)
+            for compute_resource_config in queue_config["ComputeResources"]
+        ]
+
+    def render(self):
+        """Launch the rendering of the required configuration."""
+        rendered = f"{CONFIG_HEADER}\n"
+        if self.conf_type == "gres":
+            rendered += self._as_gres()
+        else:
+            rendered += self._as_partition()
+
+        return rendered
```

Comment:
OK, this is a bit confusing at first although the code is completely fine.

In `ComputeResourceRenderer`, the `render()` method renders the node definition (compute resource) line.

In `QueueRenderer` instead the `render()` method renders the whole Slurm include file (so all the node definitions, nodeset and partition).

Could we maybe call them in two different ways? Maybe call this more general method in `QueueRenderer` something like `render_include_file()`?

Also, maybe the `_as_partition()` and `_as_gres()` method could be renamed after the portion of the Slurm config they refer to (e.g. `_for_slurm_conf()` and `_for_gres_conf`)?

Reply:
Yep. 
As discussed I'll try to rename these methods to make it more clear what they are doing and keep them consistent across both classes.

Reply:
Should be more clear now